design-proposals/network-binding-plugin/network-binding-plugin.md

```diff
@@ -257,6 +258,48 @@ VMI status and in one case to pass the data (processed) to the
 
 - DownwardAPI
 
+##### Device Info
+
+In some cases the multus status annotation
+contains [device-info](https://github.com/k8snetworkplumbingwg/device-info-spec/blob/main/SPEC.md).
+This info may be valuable for the binding plugin sidecar. For example
+`SRIOV` and `VDPA` plugins need this info.
+
+As the multus status annotation is not there yet on `virt-launcher` pod
+creation, the suggested solution is to use downwardAPI volume and mount
+it to the sidecar.
+The downwardAPI volume will track a new `virt-launcher` pod
+annotation `kubevirt.io/network-device-info-map` that will contain a
```

Comment:
I would prefer a generalization to avoid the need to create multiple values that track different annotations. We already have a good example with the existing SR-IOV one.

It should be harmes to generalize it as long we have a proper structure and a general name.
The only disadvantage is the embedded json, but we have it now anyway.

Here is a potential structure:
Annotation name: `kubevirt.io/virt-launcher-data`
Annotation value format:
```json
{
  "network": {
    "interfaces": [
      "name": "net1"
      "networkLogicalName": <logical name>
      "device-info": {...}
    ]
  }
}
```

In any case, please explicitly define the format here.

Reply:
@EdDev 
How about?
```
type VirtLauncherData struct {
	networkDeviceInfoMap map[string]*networkv1.DeviceInfo
}
```

Reply:
```go
type VirtLauncherDownwardData struct {
	Network *Network `json:"network,omitempty"`
}
type Network struct {
	Interfaces []Interface `json:"interfaces,omitempty"`
}

type Interface struct {
	Name               string `json:"name"`
	NetworkLogicalName string `json:"networkLogicalName"`
	DeviceInfo         string `json:"deviceInfo,omitempty"`
}
```

Reply:
I agree, I prefer to keep it narrow as well.
If in the future there will be a need to extend it, it should be easy, the volume will have to point to a new annotation.
The only downside is that we will have to keep the old annotation for a while to keep backward compatibility for old launchers that are already running.

Reply:
> Here we try to enable a way to pass data from virt-controller (who has access to virt-launcher pod annotation) to virt-launcher app. The suggestion to generalize the annotation may sounds appealing and future proof, it may seem like a mere annotation, but it actually become an internal API; communication channel between virt-controller and virt-launcher.

Indeed, it is a structured, stable and open for extension protocol.
It is not a path focused on one single problem that needs to be solved, it looks at the general need and defines it well.

> Currently we have similar channel, between virt-handler and launcher, using gRPC and protobuf (handler-launche-com) Unlike the `handler-launcher-com`, this annotation can easily be abused by the next dev and becoming a "go to workaround" when introducing an API change require more time and work in a tight time-frame. I am talking about cases when there is a need to pass data from virt-controller to virt-launcher and its possible using `handler-launcher-com ` / API change.

I do not accept this as an argument. Abusing a channel and protocol is always possible, it cannot be the reasoning for not establishing one.

With the same reasoning I can claim the usage of this DownwardAPI mechanism is an abuse, giving ideas to others to use the same for other means.

> Keep in mind that we struggle to keep track and apply proper review process to formal and stable APIs (hopefully it will get better with sig-api group). I am afraid this annotation will become hard to maintain and become another backward-compatibility thing to worry about.

This is true for any API change, no matter where it is placed.

> If in the future there will be a need to extend it, it should be easy, the volume will have to point to a new annotation.
> The only downside is that we will have to keep the old annotation for a while to keep backward compatibility for old launchers that are already running.

I do not think the notion of adding an API, even if internal, and admitting it may be replaced if an extension will be needed is a good direction. This implies revolution is better than evolution in a supposed-to-be stable project API.

I think the data communication channel should be formatted correctly to allow it to extend.
I disagree with narrowing its scope as argued above.
But as with the previous incident (SR-IOV DownwardAPI), I will not block the whole solution on this.